_posts/2023/03/2023-03-13-probing.md

```diff
@@ -0,0 +1,178 @@
+---
+layout: post
+date: 2023-03-13
+title: "Probing"
+author: volodya-lombrozo
+---
+
+## Overview
+
+The EO language has a convenient feature that enables the use of objects
+without the need to specify their source, location or origin. A developer can
+simply print the name of the object and that is all that is required. The
+compiler will automatically locate all the requested objects through the
+implementation of a new probing feature, which was recently added. In this short
+blog, we will attempt to explain how this feature works. Let's take a look at
+the following code snippet:
+
+```eo
+[] > app
+  org.eolang.io.stdout > @
+    org.eolang.txt.sprintf
+      "Magic number is %d"
+      4815162342
+```
+
+As you can see from the code, the purpose of this program is to print the magic
+number to the standard output. However, what is `org.eolang.io.stdout` and
+`org.eolang.io.stdout` here? They all are objects, right.
+Moreover, `org`, `org.eolang` , `org.eolang.io` are objects too.
+
+But where are they? Is it `org.eolang.io` object that contains `stdout`
+attribute, or `org.eolang.io.stdout` is the top-level object itself?
```

Comment:
@volodya-lombrozo I believe the question is not entirely correct. There is only one top-level object - `Φ`. All others are attributes of some sort. However at runtime level we do distinguish package objects. So the correct question would sound like "what part of FQN represents the package?"

Reply:
@mximp I'm afraid to add "package" term to the blog post, because we use "package" as syntax sugar only. I've already asked about it previously. On the other hand, I agree that we don't have "top-level" object either. Maybe we can name it as follows:
```
target object
free object 
non-attribute object
```
What do you think?

Reply:
@volodya-lombrozo But we do have "package objects" in runtime: namely they are the special objects which are represented not by standard `.eo` file, but by a location (or path) containing it's attributes (other objects, including other packages). So I believe package is not just a syntactic sure, but has a real representation.
All suggested names, imo, are unclear and misleading.

Reply:
@mximp Please, check that [thread](https://github.com/objectionary/eo/issues/1705#issuecomment-1399529015). I was curious about packages some time ago too. It's not my words:
> @volodya-lombrozo there are no packages in EO, but only objects -- this is the beauty/ugliness of the language. org is an object, org.eolang is an object, org.eolang.int is also an object, finally, org.eolang.int.plus is also an object. They are four object with equal rights and responsibilities. I believe that it's the job of a compiler: how to find out which object to load from where. The probing mechanism may not be perfect in its current implementation, but we must make it perfect. 

@yegor256 We need your help here.

Reply:
@volodya-lombrozo I believe there is no contradiction here. Packages are objection in terms of concept: they have attributes that are accessed the same way. But these object has special treatment in runtime (`PhPackage` class) and special syntax for definition in EO (`+package` meta). That's why you can't find any code for a package.

Reply:
@mximp I believe we have to write down the meaning of the `package` somewhere (maybe in the blog post). This ambiguity makes it difficult to reason about that term and I'm afraid to use it. When the meaning will be clear we can change current blog post. What do you think?

Reply:
@volodya-lombrozo The term "top-level object" is even more confusing than "package". How would you define it?